includes/class-indieauth-scopes.php

```diff
@@ -0,0 +1,222 @@
+ */
+class IndieAuth_Scopes {
+	private $scopes; // Stores All Registered Scopes
+
+	public function __construct() {
+		$this->scopes = array();
+		$this->register_builtin_scopes();
+		add_filter( 'map_meta_cap', array( $this, 'map_meta_cap' ), 10, 4 );
+	}
+
+	public function map_caps() {
+		return apply_filters( 'indieauth_meta_caps', array( 'publish_posts', 'delete_user', 'edit_user', 'remove_user', 'promote_user', 'delete_post', 'delete_page', 'edit_post', 'edit_page', 'read_post', 'read_page' ) );
+	}
+
+	public function map_meta_cap( $caps, $cap, $user_id, $args ) {
+		// If this is not null this is an indieauth response
+		$response = indieauth_get_response();
+		if ( ! empty( $response ) ) {
+			$scopes = indieauth_get_scopes();
+			if ( empty( $scopes ) ) {
+				return array( 'do_not_allow' );
+			}
+
+			// This check is only for certain capabilities
+			if ( ! in_array( $cap, $this->map_caps(), true ) ) {
+				if ( WP_DEBUG ) {
+					error_log( sprintf( __( 'Unknown cap: %1s', 'indieauth' ), $cap ) ); // phpcs:ignore
+				}
+				return $caps;
+			}
+			foreach ( $caps as $c ) {
+				// If the capability is not in any of the scopes then do not allow
+				if ( ! $this->has_cap( $c, $scopes ) ) {
+					return array( 'do_not_allow' );
+				}
+			}
+		}
+		return $caps;
+	}
+
+	/**
+	 * Register Scopes
+	 */
+	public function register_builtin_scopes() {
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'create',
+				__( 'Allows the application to create posts and upload to the Media Endpoint', 'indieauth' ),
+				array(
+					'edit_posts',
+					'edit_published_posts',
+					'delete_posts',
+					'publish_posts',
+					'upload_files',
+					'read',
+					'unfiltered_html',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'draft',
+				__( 'Allows the application to create draft posts only.', 'indieauth' ),
+				array(
+					'edit_posts',
+					'unfiltered_html',
+					'read',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'update',
+				__( 'Allows the application to update posts', 'indieauth' ),
+				array(
+					'edit_published_posts',
+					'edit_others_posts',
+					'read',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'delete',
+				__( 'Allows the application to delete or undelete posts', 'indieauth' ),
+				array(
+					'delete_posts',
+					'delete_published_posts',
+					'delete_others_posts',
+					'read',
+				)
+			)
+		);
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'media',
+				__( 'Allows the application to upload to the media endpoint', 'indieauth' ),
+				array(
+					'upload_files',
+					'read',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'read',
+				__( 'Allows the application read access to Microsub channels', 'indieauth' ),
+				array(
+					'read',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'follow',
```

Comment:
Shouldn't this be added by the Microsub plugin? Shouldn't the IndieAuth plugin as generic as possible?

Reply:
It is...these are all the known scopes for IndieAuth right now.

Reply:
But the comment is very specific with Microsub.